### DISCUSSION TITLE
Creating a Vercel CRON job

### DISCUSSION THREAD

**Discussion by zkSoju** [wevm/viem#368] [Question] [answered]
I'm currently trying to create a vercel CRON job that executes a transaction every 7 days, but viem relies on the NodeJs crypto, which gives me this error. I initially built the script without React in mind so I didn't use wagmi.

```
Failed to compile.

node:crypto
Module build failed: UnhandledSchemeError: Reading from "node:crypto" is not handled by plugins (Unhandled scheme)
```

Any ideas how to fix?

**Comment by fubhy** [comment]
Hi. Viem does not rely on `node:crypto`. I just tried deploying a basic serverless function to vercel and that worked. Do you have a minimal repro that you can share?

**Reply by zkSoju** [reply, marked as answer]
Solved it!

Looks like it was an issue caused by running it on the Edge, which I just copied.

```js
export const config = {
  runtime: "edge",
};
```

**Reply by fubhy** [reply]
Ok, I can confirm that it doesn't work on the edge runtime. But that's not because of `node:crypto` (viem doesn't use that), but because of `stream`, `fs` and `url` leaking into the viem from downstream deps it seems. I haven't investigated further yet but that's the error I'm seeing. Not sure if we should strive to be edge deployable @jxom @tmm?

**Reply by fubhy** [reply]
If we want that, then we'd likely have to exclude the `websocket` transport from the default export.

**Reply by fubhy** [reply]
Same with a hypothetical future ipc transport that'd likely leverage `fs`

**Reply by fubhy** [reply]
Ok, I can confirm that this is possible by just removing the websocket transport from the entrypoint. I think it's worth doing (./cc @jxom @tmm). The coldstart time benefits of edge function vs. serverless function are worth it imho. Would likely also mean that this then works on cloudflare edge & aws lambda without extra work.

**Reply by fubhy** [reply]
We might want to run a test matrix with different environments: https://vitest.dev/guide/environment.html (node, jsdom & edge)

**Comment by fubhy** [comment]
This works:

https://viem-edge-fubhy.vercel.app/api/block?block=17043877
https://github.com/fubhy/viem-edge-example

Cold start times are significantly faster than with a serverless function.

**Reply by jxom** [reply]
Interesting! We use isomorphic-ws to figure out which websocket impl to use. Can we modify it to throw on serverless?

**Reply by fubhy** [reply]
The problem appears to be that vercel's build doesn't like the existence of the import in the bundle at all. So even if we make that conditional I don't think that'd solve it. I'll continue to experiment though. Worst case, we'd simply have to provide a dedicated `viem/websocket` export and include everything related to `isomorphic-ws` in there (and only there).

**Reply by jxom** [reply]
Makes sense. If a conditional dynamic import might not work, an entrypoint might be the way to go.

**Reply by fubhy** [reply]
@jxom Scratch that, we can indeed use a dynamic import. I had simply missed one. So, we can now claim that viem also works on edge runtimes! I'll put up the PR shortly.

**Comment by fubhy** [comment]
This works now with both edge & serverless:

https://viem-edge-fubhy.vercel.app/api/edge?block=17043877
https://viem-edge-fubhy.vercel.app/api/serverless?block=17043877

https://github.com/fubhy/viem-edge-example

The PR is super simple: https://github.com/wagmi-dev/viem/pull/375

**Reply by zkSoju** [reply]
Thanks for taking a look at this! Will be giving edge a try. :)

**Reply by zkSoju** [reply]
It's odd on my end I'm still getting the same error with the `viem@0.3.2`, I am using the same code as your example in my repo.

```
You may need an additional plugin to handle "node:" URIs.
--
16:59:54.762 | at /vercel/path0/node_modules/next/dist/compiled/webpack/bundle5.js:28:395974
16:59:54.762 | at Hook.eval [as callAsync] (eval at create (/vercel/path0/node_modules/next/dist/compiled/webpack/bundle5.js:13:28771), <anonymous>:6:1)
16:59:54.763 | at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (/vercel/path0/node_modules/next/dist/compiled/webpack/bundle5.js:13:25925)
16:59:54.763 | at Object.processResource (/vercel/path0/node_modules/next/dist/compiled/webpack/bundle5.js:28:395899)
16:59:54.763 | at processResource (/vercel/path0/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:5308)
16:59:54.764 | at iteratePitchingLoaders (/vercel/path0/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:4667)
16:59:54.764 | at runLoaders (/vercel/path0/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:8590)
16:59:54.764 | at NormalModule._doBuild (/vercel/path0/node_modules/next/dist/compiled/webpack/bundle5.js:28:395761)
16:59:54.764 | at NormalModule.build (/vercel/path0/node_modules/next/dist/compiled/webpack/bundle5.js:28:397789)
16:59:54.764 | at /vercel/path0/node_modules/next/dist/compiled/webpack/bundle5.js:28:81243
16:59:54.764 |  
16:59:54.764 | Import trace for requested module:
16:59:54.764 | node:crypto
16:59:54.764 | ./node_modules/@noble/hashes/esm/cryptoNode.js
16:59:54.764 | ./node_modules/@noble/hashes/esm/utils.js
16:59:54.764 | ./node_modules/@noble/hashes/esm/sha3.js
16:59:54.764 | ./node_modules/viem/dist/esm/utils/hash/keccak256.js
16:59:54.764 | ./node_modules/viem/dist/esm/utils/hash/index.js
16:59:54.765 | ./node_modules/viem/dist/esm/utils/index.js
16:59:54.765 | ./node_modules/viem/dist/esm/chains.js


```

**Reply by fubhy** [reply]
Can you post a reproduction repo with a minimal version of your setup?

**Reply by jxom** [reply]
Hmm, looks like `@noble/hashes` uses exports to decide which crypto module to use: https://github.com/paulmillr/noble-hashes/blob/6cf0fdad677964e1bba46e3c81e9a8cedf7fd6c1/package.json#L49-L57

Wonder if it's worth opening a discussion there.

**Reply by fubhy** [reply]
But then why is my example working? I'm confused.

**Reply by jxom** [reply]
Idk lol

**Reply by fubhy** [reply]
The problem you last posted seems to be a next.js / webpack issue, not a vercel edge function issue: https://github.com/vercel/next.js/issues/28774

There's also a workaround in that issue.

**Reply by fubhy** [reply]
Also possibly related PR: https://github.com/vercel/next.js/pull/45149

**Reply by zkSoju** [reply]
> The problem you last posted seems to be a next.js / webpack issue, not a vercel edge function issue: [vercel/next.js#28774](https://github.com/vercel/next.js/issues/28774)
> 
> There's also a workaround in that issue.

I believe you're right, I'm seeing this issue when upgrading to any Next version after `13.1.6`.

**Reply by zkSoju** [reply]
https://github.com/vercel/next.js/discussions/47314#discussioncomment-5691116

I'm not sure if this is related, since it seems to have worked on that version of Next but not newer ones

**Reply by zkSoju** [reply]
https://github.com/mattddean/t3-app-router-edge-drizzle/issues/4

I think this might be the solution for the problem.

**Reply by fubhy** [reply]
Yeah we did that here too: https://github.com/jxom/viem-nextjs-edge

Sorry, should've reported back here ;-)
